octopusdeploy_framework/resource_step_template.go

```diff
@@ -0,0 +1,372 @@
+	for keys, value := range at.Properties {
+		stringProps[keys] = types.StringValue(value.Value)
+	}
+	props, dg := types.MapValue(types.StringType, stringProps)
+	resp.Append(dg...)
+	data.Properties = props
+
+	// Packages
+	pkgs, dg := convertStepTemplateToPackageAttributes(at.Packages)
+	resp.Append(dg...)
+	data.Packages = pkgs
+
+	return resp
+}
+
+func mapStepTemplateResourceModelToActionTemplate(ctx context.Context, data schemas.StepTemplateTypeResourceModel) (*actiontemplates.ActionTemplate, diag.Diagnostics) {
+	resp := diag.Diagnostics{}
+	at := actiontemplates.NewActionTemplate(data.Name.ValueString(), data.ActionType.ValueString())
+
+	at.SpaceID = data.SpaceID.ValueString()
+	at.Description = data.Description.ValueString()
+	if !data.CommunityActionTemplateId.IsNull() {
+		at.CommunityActionTemplateID = data.CommunityActionTemplateId.ValueString()
+	}
+
+	pkgs := make([]schemas.StepTemplatePackageType, 0, len(data.Packages.Elements()))
+	resp.Append(data.Packages.ElementsAs(ctx, &pkgs, false)...)
+	if resp.HasError() {
+		return at, resp
+	}
+
+	props := make(map[string]types.String, len(data.Properties.Elements()))
+	resp.Append(data.Properties.ElementsAs(ctx, &props, false)...)
+	if resp.HasError() {
+		return at, resp
+	}
+
+	params := make([]schemas.StepTemplateParameterType, 0, len(data.Parameters.Elements()))
+	resp.Append(data.Parameters.ElementsAs(ctx, &params, false)...)
+	if resp.HasError() {
+		return at, resp
+	}
+
+	if len(props) > 0 {
+		templateProps := make(map[string]core.PropertyValue, len(props))
+		for key, val := range props {
+			templateProps[key] = core.NewPropertyValue(val.ValueString(), false)
+		}
+		at.Properties = templateProps
+	} else {
+		at.Properties = make(map[string]core.PropertyValue)
+	}
+
+	at.Packages = make([]packages.PackageReference, len(pkgs))
+	if len(pkgs) > 0 {
+		for i, val := range pkgs {
+			pkgProps := make(map[string]string, len(val.Properties.Attributes()))
+			for key, prop := range val.Properties.Attributes() {
+				if prop.Type(ctx) == types.StringType {
+					pkgProps[key] = prop.(types.String).ValueString()
+				} else {
+					// We should not get this error unless we add a field to package properties in the schema that is not a string
+					resp.AddError("Unexpected value type in package properties.",
+						fmt.Sprintf("Expected [%s] to have value of string but got [%s].", key, prop.String()))
+				}
+			}
+			if resp.HasError() {
+				return at, resp
+			}
+			pkgRef := packages.PackageReference{
+				AcquisitionLocation: val.AcquisitionLocation.ValueString(),
+				FeedID:              val.FeedID.ValueString(),
+				Properties:          pkgProps,
+				Name:                val.Name.ValueString(),
+				PackageID:           val.PackageID.ValueString(),
+			}
+			pkgRef.ID = val.ID.ValueString()
+			at.Packages[i] = pkgRef
+		}
+	}
+
+	at.Parameters = make([]actiontemplates.ActionTemplateParameter, len(params))
+	if len(params) > 0 {
```

Comment:
nit question: Is the reason for `if len(params) > 0` is to avoid allocation of `paramIDMap` when parameters are empty?

Reply:
essentially, yes. To be honest I think this is leftover from the initial implementation before I did a refactor. Going to leave as is but I am not normally this memory concerned.